File: C++/003.cpp
#include <iostream>

long long largestPrime(long a)
{
    long long largestFactor = -1 ;

    if(a%2==0)
    {
        largestFactor = 2;
        a/=2;
    }

    for (long long i = 3; i * i <= a; i += 2) {
        while (a % i == 0) {
            largestFactor = i;
            a /= i;
        }
    }
    
    if (a > 2) {
        largestFactor = a;
    }

    return largestFactor;

}

int main() {
    long long number = 600851475143;
    std::cout << "The largest prime factor of " << number << " is " << largestPrime(number) << std::endl;
    return 0;
}

File: Rust/003.rs
fn largest_prime_factor(mut n: u64) -> u64 {
    let mut largest_factor = 0;

    // Handle factor 2
    while n % 2 == 0 {
        largest_factor = 2;
        n /= 2;
    }

    // Handle odd factors
    let mut i = 3;
    while i * i <= n {
        while n % i == 0 {
            largest_factor = i;
            n /= i;
        }
        i += 2;
    }

    // If n is a prime number greater than 2
    if n > 2 {
        largest_factor = n;
    }

    largest_factor
}

fn main() {
    let number = 600851475143;
    println!("The largest prime factor of {} is {}", number, largest_prime_factor(number));
}